internal/core/metadata/v2/controller/http/deviceservice_test.go

```diff
@@ -0,0 +1,169 @@
+			Name:           testDeviceServiceName,
+			Description:    TestDescription,
+			Labels:         testDeviceServiceLabels,
+			AdminState:     models.Unlocked,
+			OperatingState: models.Enabled,
+			BaseAddress:    testBaseAddress,
+		},
+	}
+
+	return testAddDeviceServiceReq
+}
+
+func buildTestDBClient(dsModel models.DeviceService, errKind errors.ErrKind, errorMessage string) *dbMock.DBClient {
+	dbClientMock := &dbMock.DBClient{}
+	if len(errKind) > 0 {
+		err := errors.NewCommonEdgeX(errKind, errorMessage, nil)
+		dbClientMock.On("AddDeviceService", dsModel).Return(dsModel, err)
+	} else {
+		dbClientMock.On("AddDeviceService", dsModel).Return(dsModel, nil)
+	}
+	return dbClientMock
+}
+
+func TestAddDeviceService(t *testing.T) {
+	validReq := buildTestDeviceServiceRequest()
+	dsModels := requests.AddDeviceServiceReqToDeviceServiceModels([]requests.AddDeviceServiceRequest{validReq})
+	expectedRequestId := ExampleUUID
+	normalMessage := fmt.Sprintf("Add device service %s successfully", testDeviceServiceName)
+	duplicateServiceNameMessage := fmt.Sprintf("device service %s already exists", testDeviceServiceName)
+
+	reqWithNoID := validReq
+	reqWithNoID.RequestID = ""
+	reqWithInvalidId := validReq
+	reqWithInvalidId.RequestID = "InvalidUUID"
+	reqWithNoName := validReq
+	reqWithNoName.Service.Name = ""
+
+	tests := []struct {
+		name                   string
+		isValidRequest         bool
+		dbClientMock           *dbMock.DBClient
+		Request                []requests.AddDeviceServiceRequest
+		expectedHttpStatusCode int
+		expectedMessage        string
+	}{
+		{
+			"Request Normal",
+			true,
+			buildTestDBClient(dsModels[0], "", ""),
+			[]requests.AddDeviceServiceRequest{validReq},
+			http.StatusCreated,
+			normalMessage,
+		},
+		{
+			"Request without requestId",
+			true,
+			buildTestDBClient(dsModels[0], "", ""),
+			[]requests.AddDeviceServiceRequest{reqWithNoID},
+			http.StatusCreated,
+			normalMessage,
+		},
+		{
+			"Request with duplicate service name",
+			true,
+			buildTestDBClient(dsModels[0], errors.KindDuplicateName, duplicateServiceNameMessage),
+			[]requests.AddDeviceServiceRequest{validReq},
+			http.StatusConflict,
+			duplicateServiceNameMessage,
+		},
+		{
+			"Request with invalid requestId",
+			false,
+			buildTestDBClient(dsModels[0], "", ""),
+			[]requests.AddDeviceServiceRequest{reqWithInvalidId},
+			http.StatusBadRequest,
+			"",
+		},
+		{
+			"Request without service name",
+			false,
+			buildTestDBClient(dsModels[0], "", ""),
+			[]requests.AddDeviceServiceRequest{reqWithNoName},
+			http.StatusBadRequest,
+			"",
+		},
+	}
+	for _, testCase := range tests {
+		t.Run(testCase.name, func(t *testing.T) {
+
+			dic := mockDic()
+			dic.Update(di.ServiceConstructorMap{
+				v2MetadataContainer.DBClientInterfaceName: func(get di.Get) interface{} {
+					return testCase.dbClientMock
+				},
+			})
+
+			controller := NewDeviceServiceController(dic)
+			assert.NotNil(t, controller)
```

Comment:
Should be require.NotNil() to avoid panic on line 48

Reply:
Thanks for the catch!  Updated.